libs/ballot-interpreter/src/hmpb-rust/image_utils.rs

```diff
@@ -278,6 +285,93 @@ pub fn find_scanned_document_inset(image: &GrayImage, threshold: u8) -> Option<I
+    const MAX_WHITE_GAP_PIXELS: PixelUnit = 15;
+    const BORDER_COLUMNS_TO_EXCLUDE: PixelUnit = 5;
+
+    // Look at each column of pixels in the image (ignoring
+    // BORDER_COLUMNS_TO_EXCLUDE on either side), where a "column" is two pixels
+    // wide. If more than PERCENT_BLACK_PIXELS_IN_STREAK of the rows in the
+    // column have a black pixel, it might be a streak. Filter out streaks that
+    // have gaps of white that are greater than MAX_WHITE_GAP_PIXELS, since
+    // these are probably printed features, not streaks. This relies on the
+    // invariant that there are no printed features that span the entire page
+    // top to bottom without a gap greater than MAX_WHITE_GAP_PIXELS.
+
+    let (width, height) = image.dimensions();
+    let binarized_columns =
+        (BORDER_COLUMNS_TO_EXCLUDE - 1..width - BORDER_COLUMNS_TO_EXCLUDE).map(|x| {
+            let binarized_column = (0..height)
+                .map(|y| {
+                    [image.get_pixel(x, y), image.get_pixel(x + 1, y)]
+                        .iter()
+                        .any(|pixel| pixel[0] <= threshold)
+                })
+                .collect::<Vec<_>>();
+            (x, binarized_column)
+        });
+
+    let streak_columns = binarized_columns
+        .filter_map(|(x, column)| {
+            let num_black_pixels = column.iter().filter(|is_black| **is_black).count();
+            let percent_black_pixels = num_black_pixels as f32 / height as f32;
+            if percent_black_pixels >= PERCENT_BLACK_PIXELS_IN_STREAK {
+                Some((x, column, percent_black_pixels))
+            } else {
+                None
+            }
+        })
+        .filter_map(|(x, column, percent_black_pixels)| {
+            let mut longest_white_gap_length = 0;
+            let mut current_gap_length = 0;
+            for is_black in column {
+                if is_black {
+                    longest_white_gap_length = longest_white_gap_length.max(current_gap_length);
+                    current_gap_length = 0;
+                } else {
+                    current_gap_length += 1;
+                }
+            }
+            longest_white_gap_length = longest_white_gap_length.max(current_gap_length);
+            if longest_white_gap_length <= MAX_WHITE_GAP_PIXELS {
+                Some((x, percent_black_pixels, longest_white_gap_length))
+            } else {
+                None
+            }
+        });
+
+    // If there are adjacent streak columns, just pick one to represent the streak.
+    let streaks = streak_columns.fold(vec![], |mut combined_streaks, column| {
+        if let Some(last_column) = combined_streaks.pop() {
+            let (last_x, _, _) = last_column;
+            let (x, _, _) = column;
+            if x - last_x != 1 {
+                combined_streaks.push(last_column);
+            }
+        }
+        combined_streaks.push(column);
+        combined_streaks
+    });
```

Comment:
Minor suggestion for some improvements:

```suggestion
    let streak_columns = binarized_columns.filter_map(|(x, column)| {
        let num_black_pixels = column.iter().filter(|is_black| **is_black).count();
        let percent_black_pixels = num_black_pixels as f32 / height as f32;
        if percent_black_pixels < PERCENT_BLACK_PIXELS_IN_STREAK {
            return None;
        }

        let longest_white_gap_length = column
            .into_iter()
            .group_by(|is_black| *is_black)
            .into_iter()
            .filter(|(is_black, _)| !*is_black)
            .map(|(_, white_gap_pixels)| white_gap_pixels.count() as PixelUnit)
            .max()
            .unwrap_or(0);

        if longest_white_gap_length <= MAX_WHITE_GAP_PIXELS {
            Some((x, percent_black_pixels, longest_white_gap_length))
        } else {
            None
        }
    });

    // If there are adjacent streak columns, just pick one to represent the streak.
    let streaks = streak_columns
        .coalesce(|(x1, p1, l1), (x2, p2, l2)| {
            if x2 - x1 == 1 {
                Ok((x2, p2, l2))
            } else {
                Err(((x1, p1, l1), (x2, p2, l2)))
            }
        })
        .collect::<Vec<_>>();
```

Reply:
Nice! I had something similar to your approach for calculating the gap length, but it was pretty slow. Though I didn't know about `group_by`, so I had implemented my own version that required materializing the iterator into a vector, which slowed things down a lot. I didn't really expect that `group_by` would be the name for this logic, since I generally associate it with grouping the entire set, not just consecutive runs. Anyway, this version seems fast, yay!

`coalesce` is also exactly what I was looking for but couldn't find, thanks!

Reply:
I hear you on `group_by`–I had the same thought. 😬